clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -1569,6 +1576,27 @@ class KernelObjVisitor {
     else if (ParamTy->isStructureOrClassType()) {
       if (KP_FOR_EACH(handleStructType, Param, ParamTy)) {
         CXXRecordDecl *RD = ParamTy->getAsCXXRecordDecl();
+        llvm::SmallVector<const CXXRecordDecl *, 8> WorkList;
```

Comment:
Hmm, this is not exactly what I meant about using visitors infrastructure. What I meant is that we should extend `SyclKernelFieldChecker` class, so it checks bases. It has methods that accept `CXXBaseSpecifier` AST node which represent base specifier. Each `CXXBaseSpecifier` knows whether it is virtual. I think `SyclKernelFieldChecker` should be already recursively visiting each struct it meets. We just need to diagnose from some of method of SyclKernelFieldChecker` that accepts `CXXBaseSpecifier` . Because this is not even specific for free function kernels, you don't even need to invent a method specific for free function kernels and accepting `CXXBaseSpecifier`.